src/utils/common.ts

```diff
@@ -290,33 +292,57 @@ export async function estimateTotalGasRequiredByUnsignedTransaction(
 }
 
 /**
- * Create an unsigned transaction of a fillRelay contract call
- * @param spokePool The specific spokepool that will populate this tx
- * @param destinationTokenAddress A valid ERC20 token (system-wide default is UDSC)
- * @param simulatedRelayerAddress The relayer address that relays this transaction
- * @returns A populated (but unsigned) transaction that can be signed/sent or used for estimating gas costs
+ * Create an unsigned transaction to fill a relay. This function is used to simulate the gas cost of filling a relay.
+ * @param spokePool A valid SpokePool contract instance
+ * @param fillToSimulate The fill that this function will use to populate the unsigned transaction
+ * @returns An unsigned transaction that can be used to simulate the gas cost of filling a relay
  */
-export async function createUnsignedFillRelayTransaction(
+export async function createUnsignedFillRelayTransactionFromFill(
   spokePool: SpokePool,
-  destinationTokenAddress: string,
-  simulatedRelayerAddress: string
+  fillToSimulate: Fill
 ): Promise<PopulatedTransaction> {
-  // Populate and return an unsigned tx as per the given spoke pool
-  // NOTE: 0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B is a dummy address
-  return await spokePool.populateTransaction.fillRelay(
-    "0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B",
-    "0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B",
-    destinationTokenAddress,
-    "10",
-    "10",
-    "1",
-    "1",
-    "1",
-    "1",
-    "1",
-    [],
+  const {
+    message,
+    recipient: recipientAddress,
+    relayer: relayerAddress,
+    amount: amountToRelay,
+    depositor: depositorAddress,
+  } = fillToSimulate;
+  const provider = spokePool.provider;
+
+  const isMessageEmpty = message.length === 0 || message === "0x";
+  // Only continue sanity checks if the message is not empty
+  if (!isMessageEmpty) {
+    const isRecipientAnAddress = ethers.utils.isAddress(recipientAddress);
+    const isRelayerAnAddress = ethers.utils.isAddress(relayerAddress);
+    if (!isRecipientAnAddress || !isRelayerAnAddress) {
+      throw new Error("Could not simulate message fill. Recipient address or relayer address is not a valid address");
+    }
+    const [isRecipientAContract, relayerBalanceOfToken] = await Promise.all([
+      isContractAddress(recipientAddress, provider),
+      provider.getBalance(fillToSimulate.destinationToken),
+    ]);
+    if (!isRecipientAContract) {
+      throw new Error("Could not simulate message fill. Recipient address is not a contract address");
+    }
+    if (toBN(relayerBalanceOfToken).lt(toBN(amountToRelay))) {
```

Comment:
In this case, you might want to instead reset `amountToRelay = relayerBalanceOfToken` so that the constructed fill still is a "full fill" but it can be simulated.

Reply:
Right, but what if the contract that implements the `handleAcrossMessage` is specifically looking at `amount` and `tokenSent` as a conditional path in their logic? Couldn't changing the request intrinsically affect the gas cost estimation?

Reply:
Ah shoot that's a fair point. Thoughts on doing what we did with the `amountTooLow` variable? I.e. we pass another optional query flag to the request like `ignoreMessageFlowFillWarning` and we could relax that restraint conditionally?

Reply:
hmmm that could work. Do you think the parameter/interface is too big and we should keep the calculator simpler? This would mean foisting off the responsibility to the client to determine whether to catch the result and then resubmit with a lower amount? I'm not sure how to lean here, what do you think?

Reply:
Right - I'd posit that any developer using this specific endpoint would rather their incorrect assumption of the system return a loud error than for it to return a quote

Reply:
although, i can see an extra flag being useful so that the caller doesn't have to call into the SDK twice and make 2x RPC calls. The alternative of checking the simulated EOA balance and then calling into the SDK results also in 2x RPC calls, but this is unavoidable, as either the caller checks the EOA balance before calling into the SDK or the SDK will do it inevitably itself.

Reply:
I'd like to leave this as-is given that this is the best way to effectively estimate gas and ensure that we're not potentially making changes that affect the output of the arbitrary `handleAcrossMessage` contract

Reply:
So we're going to leave this as is and make caller to catch the error and readjust the `amountToRelay` they pass in? 

Just to confirm